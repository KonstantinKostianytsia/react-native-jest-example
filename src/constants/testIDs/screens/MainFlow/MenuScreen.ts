const screenName = 'MenuScreen';

export const MenuScreenTestIDs = {
  mainButton: `${screenName}.Button.MainButton`,
};
