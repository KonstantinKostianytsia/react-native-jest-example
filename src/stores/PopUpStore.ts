import {makeAutoObservable} from 'mobx';

export class PopUpStore {
  isShown: boolean = false;

  constructor() {
    makeAutoObservable(this);
  }

  showPopUp() {
    this.isShown = true;
  }

  closePopUp() {
    this.isShown = false;
  }
}
