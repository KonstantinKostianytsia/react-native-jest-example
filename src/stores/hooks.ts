import rootStore, {RootStore} from 'stores';

export const useStore = (): RootStore => rootStore;

export const usePopUpStore = () => useStore().popUpStore;
