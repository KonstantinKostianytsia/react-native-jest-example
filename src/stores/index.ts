import {PopUpStore} from './PopUpStore';

export class RootStore {
  popUpStore: PopUpStore;

  constructor() {
    this.popUpStore = new PopUpStore();
  }
}

const rootStore = new RootStore();

export default rootStore;
