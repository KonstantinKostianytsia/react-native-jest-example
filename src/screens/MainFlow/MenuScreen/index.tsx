import React from 'react';

import BackgroundView from 'components/atoms/BackgroundView';
import {PrimaryH1Label} from 'components/atoms/Labels';
import {usePopUpStore} from 'stores/hooks';
import {Observer} from 'mobx-react';
import {CustomButton} from 'components/atoms/CustomButton';
import {MenuScreenTestIDs} from 'constants/testIDs/screens/MainFlow/MenuScreen';

const MenuScreen = () => {
  const popUpStore = usePopUpStore();
  return (
    <Observer>
      {() => (
        <BackgroundView>
          <CustomButton
            testID={MenuScreenTestIDs.mainButton}
            title="Toogle pop up"
            onPress={() =>
              popUpStore.isShown
                ? popUpStore.closePopUp()
                : popUpStore.showPopUp()
            }
          />
          {popUpStore.isShown && (
            <PrimaryH1Label testID="PopUpLabel">Pop Up is shown</PrimaryH1Label>
          )}
        </BackgroundView>
      )}
    </Observer>
  );
};

export default MenuScreen;
