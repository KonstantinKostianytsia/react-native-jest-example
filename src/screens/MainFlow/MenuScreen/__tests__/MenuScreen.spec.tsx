import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';
import MenuScreen from 'screens/MainFlow/MenuScreen';
import {mockedPopUpStore} from 'stores/__test__/__mock__/PopUpStore.mock';
import {usePopUpStore} from 'stores/hooks';
import {MenuScreenTestIDs} from 'constants/testIDs/screens/MainFlow/MenuScreen';

describe('MenuScreen:', () => {
  it('Screen renders correctly', () => {
    const screen = render(<MenuScreen />);
    expect(screen.toJSON()).toMatchSnapshot();
  });
  it('Label appears on click on `toogle button`', async () => {
    const screen = render(<MenuScreen />);
    // Fire event
    const toggleButton = screen.getByTestId(MenuScreenTestIDs.mainButton);
    fireEvent.press(toggleButton);

    expect(mockedPopUpStore.showPopUp).toHaveBeenCalled();
  });

  it('Label appears if `isShown` equals true', () => {
    (usePopUpStore as jest.Mock).mockImplementationOnce(() => ({
      ...mockedPopUpStore,
      isShown: true,
    }));

    const screen = render(<MenuScreen />);
    const popUpLabel = screen.queryAllByTestId('PopUpLabel');
    expect(popUpLabel.length).toBe(1);
  });
});
