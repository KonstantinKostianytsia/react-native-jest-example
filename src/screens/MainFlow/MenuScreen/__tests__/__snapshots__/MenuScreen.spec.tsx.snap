// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuScreen: Screen renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="MenuScreen.Button.MainButton"
  >
    <Text
      testID="ButtonTitle"
    >
      Toogle pop up
    </Text>
  </View>
</View>
`;
